samples/sync_object.html

```diff
@@ -0,0 +1,422 @@
+        }
+
+        function emitParticles() {
+            time = Date.now();
+
+            var programEmit = programs[PROGRAM_EMIT];
+
+            // Toggle source and destination VBO
+            var sourceVBO = particleVBOs[currentSourceIdx];
+            var sourceVAO = particleVAOs[currentSourceIdx];
+
+            var destinationVBO = particleVBOs[(currentSourceIdx + 1) % 2];
+
+            gl.bindTransformFeedback(gl.TRANSFORM_FEEDBACK, transformFeedback);
+
+            gl.useProgram(programEmit);
+
+            setupVertexAttributes(sourceVAO, sourceVBO);
+
+            // Set transform feedback buffer
+            gl.bindBufferBase(gl.TRANSFORM_FEEDBACK_BUFFER, 0, destinationVBO[Particle.POSITION]);
+            gl.bindBufferBase(gl.TRANSFORM_FEEDBACK_BUFFER, 1, destinationVBO[Particle.VELOCITY]);
+            gl.bindBufferBase(gl.TRANSFORM_FEEDBACK_BUFFER, 2, destinationVBO[Particle.SPAWN_TIME]);
+            gl.bindBufferBase(gl.TRANSFORM_FEEDBACK_BUFFER, 3, destinationVBO[Particle.LIFE_TIME]);
+
+            // Turn off rasterization - we are not drawing
+            gl.enable(gl.RASTERIZER_DISCARD);
+
+            // Set uniforms
+            gl.uniform1f(emitTimeLocation, time);
+            gl.uniform2f(emitAccelerationLocation, 0.0, ACCELERATION);
+
+            // Emit particles using transform feedback
+            gl.beginTransformFeedback(gl.POINTS);
+            gl.drawArrays(gl.POINTS, 0, NUM_PARTICLES);
+            gl.endTransformFeedback();
+
+            // Create a sync object to ensure transform feedback results
+            // are completed before the draw that uses them.
+            // Sync object can't be reused so we create a new one for each synchronization operation.
+            // Not recommended to use gl.finish unless neccessary for synchronizing all your operation as it could severely reduce the GPU performance. Instead use sync object as you can wait for partial completion of the GL command stream.
+            emitSync = gl.fenceSync(gl.SYNC_GPU_COMMANDS_COMPLETE, 0);
+
+            // Restore state
+            gl.disable(gl.RASTERIZER_DISCARD);
+            gl.useProgram(null);
+            gl.bindBuffer(gl.ARRAY_BUFFER, null);
+            gl.bindBufferBase(gl.TRANSFORM_FEEDBACK_BUFFER, 0, null);
+            gl.bindBufferBase(gl.TRANSFORM_FEEDBACK_BUFFER, 1, null);
+            gl.bindBufferBase(gl.TRANSFORM_FEEDBACK_BUFFER, 2, null);
+            gl.bindBufferBase(gl.TRANSFORM_FEEDBACK_BUFFER, 3, null);
+            gl.bindTransformFeedback(gl.TRANSFORM_FEEDBACK, null);
+
+            // Ping pong the buffers
+            currentSourceIdx = (currentSourceIdx + 1) % 2;
+        }
+
+        function render() {
+            // Spawn particles
+            emitParticles();
+
+            time = Date.now();
+
+            // Block the GL server until transform feedback results are completed
+            gl.waitSync(emitSync, 0, gl.TIMEOUT_IGNORED);

```

Comment:
It isn't actually necessary to issue or wait on these sync objects. It would be necessary if multiple WebGL contexts were in use, and resources were being shared between them -- but the sharing semantics are difficult enough to make portable that to date WebGL only supports independent contexts. In a single context, all commands are issued serially, and the OpenGL implementation is responsible for ensuring that the transform feedback pass is complete by the time the render pass is run. See https://github.com/toji/webgl2-particles-2 for example.

While I understand that you're trying to demonstrate use of sync objects, I'm concerned that naive developers may copy your sample and add useless sync objects to their applications, possibly impacting performance.

Would you consider repurposing this sample to just demonstrate transform feedback, without the use of sync objects?